File: backend/ml/train_model.py
"""End-to-end training pipeline — the reference scikit-learn implementation.

Pipeline:
   1. Load dataset/disease_symptoms.csv
   2. Validate the dataset (columns, coverage, ranges, missing values)
   3. Normalize symptom names
   4. Generate binary training instances from the documented conditional
      probability table (seeded, reproducible)
   5. Encode disease labels (LabelEncoder)
   6. Stratified 80/20 train/test split
   7. Train a Random Forest Classifier
   8. Evaluate: accuracy, precision, recall, F1, confusion matrix
   9. Save disease_model.pkl, label_encoder.pkl, feature_columns.pkl and
      metrics.json into backend/model/

Usage:  python backend/ml/train_model.py        (from the project root)
"""

from __future__ import annotations

import datetime
import json
from pathlib import Path

import joblib
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

from evaluate_model import evaluate_model
from preprocessing import (
    FEATURE_COLUMNS,
    GENERATION_SEED,
    INSTANCES_PER_DISEASE,
    NOISE,
    generate_instances,
    load_probability_table,
)

PROJECT_ROOT = Path(__file__).resolve().parents[2]
DATASET_PATH = PROJECT_ROOT / "dataset" / "disease_symptoms.csv"
MODEL_DIR = PROJECT_ROOT / "backend" / "model"

TEST_FRACTION = 0.2
SPLIT_SEED = 7
RF_PARAMS = {
    "n_estimators": 60,
    "max_depth": 10,
    "min_samples_leaf": 5,
    "random_state": 42,
    "class_weight": None,
}


def main() -> dict:
    print("=" * 60)
    print("  AI Disease Prediction — Training Pipeline (Python)")
    print("=" * 60)

    print(f"[1/7] Loading dataset: {DATASET_PATH}")
    df = load_probability_table(DATASET_PATH)
    print(
        f"      → {len(df)} probability rows, "
        f"{df['disease'].nunique()} diseases, {len(FEATURE_COLUMNS)} features."
    )

    print("[2/7] Generating binary instances (seeded sampling)")
    features, labels, classes = generate_instances(df)
    print(f"      → {len(features)} instances, {features.shape[1]} binary features.")

    print(f"[3/7] Stratified split ({int((1 - TEST_FRACTION) * 100)}%/{int(TEST_FRACTION * 100)}%)")
    X_train, X_test, y_train, y_test = train_test_split(
        features, labels, test_size=TEST_FRACTION, random_state=SPLIT_SEED,
        stratify=labels,
    )
    print(f"      → train: {len(X_train)}, test: {len(X_test)}")

    print("[4/7] Encoding labels")
    encoder = LabelEncoder()
    y_train_enc = encoder.fit_transform(y_train)
    y_test_enc = encoder.transform(y_test)
    # Encoder classes are sorted alphabetically — align our class list with it.
    classes = list(encoder.classes_)

    print(f"[5/7] Training Random Forest {RF_PARAMS}")
    clf = RandomForestClassifier(**RF_PARAMS)
    clf.fit(X_train, y_train_enc)

    print("[6/7] Evaluating on the held-out test set")
    y_pred = clf.predict(X_test)
    report = evaluate_model(y_test_enc, y_pred, classes)
    print(f"      → Accuracy : {report['accuracy']:.1%}")
    print(f"      → Precision: {report['precision']:.1%} (macro)")
    print(f"      → Recall   : {report['recall']:.1%} (macro)")
    print(f"      → F1-score : {report['f1']:.1%} (macro)")
    print(report["classification_report"])

    print(f"[7/7] Saving artifacts → {MODEL_DIR}")
    MODEL_DIR.mkdir(parents=True, exist_ok=True)
    joblib.dump(clf, MODEL_DIR / "disease_model.pkl")
    joblib.dump(encoder, MODEL_DIR / "label_encoder.pkl")
    joblib.dump(FEATURE_COLUMNS, MODEL_DIR / "feature_columns.pkl")
    metrics = {
        "algorithm": "Random Forest",
        "generated_at": datetime.datetime.now().isoformat(),
        "n_diseases": len(classes),
        "n_symptoms": len(FEATURE_COLUMNS),
        "train_samples": int(len(X_train)),
        "test_samples": int(len(X_test)),
        "dataset": {
            "instances_per_disease": INSTANCES_PER_DISEASE,
            "noise": NOISE,
            "seed": GENERATION_SEED,
            "total_instances": int(len(features)),
            "source": str(DATASET_PATH),
        },
        "params": RF_PARAMS,
        "classes": classes,
        "feature_columns": FEATURE_COLUMNS,
        "feature_importance": [
            round(float(v), 6) for v in clf.feature_importances_
        ],
        **{k: v for k, v in report.items() if k != "classification_report"},
    }
    (MODEL_DIR / "metrics.json").write_text(json.dumps(metrics, indent=2), encoding="utf-8")
    print("      → disease_model.pkl, label_encoder.pkl, feature_columns.pkl, metrics.json written.")

    print("=" * 60)
    print(f"  Training complete. Test accuracy: {report['accuracy']:.1%}")
    print("=" * 60)
    return metrics


if __name__ == "__main__":
    main()

File: ml/train_model.ts
/**
 * End-to-end training pipeline — mirrors backend/ml/train_model.py.
 *
 * Pipeline:
 *   1. Load dataset/disease_symptoms.csv
 *   2. Validate the dataset (coverage, ranges, missing values)
 *   3. Normalize symptom names
 *   4. Generate binary training instances from the documented conditional
 *      probability table (seeded, reproducible)
 *   5. Encode disease labels
 *   6. Stratified train/test split (80/20)
 *   7. Train a Random Forest classifier (from scratch, see randomForest.ts)
 *   8. Evaluate on the held-out test set (accuracy / precision / recall / F1 /
 *      confusion matrix)
 *   9. Write the serialized model + evaluation metadata to
 *      src/convex/ml_model.ts (the artifact used by the Convex backend at
 *      runtime — no hard-coded predictions anywhere).
 *
 * Run from the project root:  bun ml/train_model.ts   (or: bun run ml:train)
 */

import { writeFileSync } from "node:fs";
import type { ModelInfo, RandomForestModel } from "../src/convex/ml/types";
import { FEATURE_COLUMNS } from "../src/convex/ml/types";
import { evaluate } from "./evaluate_model";
import {
  generateInstances,
  loadProbabilityTable,
  trainTestSplit,
  validateTable,
} from "./preprocessing";
import { predictClass, trainRandomForest } from "../src/convex/ml/randomForest";

const DATASET_PATH = "dataset/disease_symptoms.csv";
const OUTPUT_PATH = "src/convex/ml_model.ts";

const INSTANCES_PER_DISEASE = 300;
const NOISE = 0.03;
const GENERATION_SEED = 42;
const SPLIT_SEED = 7;
const TEST_FRACTION = 0.2;

const PARAMS = { nEstimators: 60, maxDepth: 10, minSamplesLeaf: 5, randomSeed: 42 };

function pct(x: number): string {
  return `${(x * 100).toFixed(1)}%`;
}

function main(): void {
  console.log("═══════════════════════════════════════════════════");
  console.log("  AI Disease Prediction — Model Training Pipeline");
  console.log("═══════════════════════════════════════════════════\n");

  // 1–3. Load, validate, normalize.
  console.log("[1/7] Loading dataset:", DATASET_PATH);
  const rows = loadProbabilityTable(DATASET_PATH);
  validateTable(rows);
  const diseases = [...new Set(rows.map((r) => r.disease))].sort();
  console.log(`      → ${rows.length} probability rows, ${diseases.length} diseases, ${FEATURE_COLUMNS.length} symptom features.`);
  console.log("      → Dataset validation passed (no missing values, all probabilities in [0,1]).\n");

  // 4–5. Generate instances + encode labels.
  console.log("[2/7] Generating binary training instances (documented sampling, seed", `${GENERATION_SEED}).`);
  const dataset = generateInstances(rows, {
    instancesPerDisease: INSTANCES_PER_DISEASE,
    noise: NOISE,
    seed: GENERATION_SEED,
  });
  console.log(`      → ${dataset.features.length} instances (${INSTANCES_PER_DISEASE} per disease), ${dataset.features[0].length} binary features.\n`);

  // 6. Stratified split.
  console.log("[3/7] Stratified train/test split (", `${(1 - TEST_FRACTION) * 100}% / ${TEST_FRACTION * 100}%).`);
  const { trainFeatures, trainLabels, testFeatures, testLabels } = trainTestSplit(
    dataset,
    TEST_FRACTION,
    SPLIT_SEED,
  );
  console.log(`      → train: ${trainFeatures.length}, test: ${testFeatures.length}\n`);

  // 7. Train.
  console.log("[4/7] Training Random Forest (", `${PARAMS.nEstimators} trees, max_depth=${PARAMS.maxDepth}, min_samples_leaf=${PARAMS.minSamplesLeaf}, seed=${PARAMS.randomSeed}).`);
  const started = Date.now();
  const model = trainRandomForest(trainFeatures, trainLabels, dataset.classes, [...FEATURE_COLUMNS], PARAMS);
  const elapsedMs = Date.now() - started;
  const totalNodes = model.trees.reduce((acc, t) => acc + t.nodes.length / 5, 0);
  console.log(`      → trained in ${(elapsedMs / 1000).toFixed(2)}s (${model.trees.length} trees, ${totalNodes} nodes).\n`);

  // 8. Evaluate.
  console.log("[5/7] Evaluating on held-out test set...");
  const predictions = testFeatures.map((f) => predictClass(model, f));
  const evalResult = evaluate(testLabels, predictions, dataset.classes);
  console.log(`      → Accuracy : ${pct(evalResult.accuracy)}`);
  console.log(`      → Precision: ${pct(evalResult.precision)} (macro)`);
  console.log(`      → Recall   : ${pct(evalResult.recall)} (macro)`);
  console.log(`      → F1-score : ${pct(evalResult.f1)} (macro)\n`);

  console.log("[6/7] Per-class report:");
  console.log("      ┌────────────────────────┬──────────┬──────────┬──────────┬─────────┐");
  console.log("      │ disease                 │ precision│ recall   │ f1       │ support │");
  console.log("      ├────────────────────────┼──────────┼──────────┼──────────┼─────────┤");
  for (const row of evalResult.perClass) {
    console.log(
      `      │ ${row.disease.padEnd(23)} │ ${pct(row.precision).padStart(7)} │ ${pct(row.recall).padStart(7)} │ ${pct(row.f1).padStart(7)} │ ${String(row.support).padStart(7)} │`,
    );
  }
  console.log("      └────────────────────────┴──────────┴──────────┴──────────┴─────────┘\n");

  // 9. Serialize.
  console.log("[7/7] Writing model artifact →", OUTPUT_PATH);
  const info: ModelInfo = {
    algorithm: "Random Forest",
    generatedAt: new Date().toISOString(),
    nDiseases: dataset.classes.length,
    nSymptoms: FEATURE_COLUMNS.length,
    trainSamples: trainFeatures.length,
    testSamples: testFeatures.length,
    dataset: {
      instancesPerDisease: INSTANCES_PER_DISEASE,
      noise: NOISE,
      seed: GENERATION_SEED,
      totalInstances: dataset.features.length,
      source: "dataset/disease_symptoms.csv (documented conditional-probability table; educational demonstration)",
    },
    params: { ...PARAMS, featureSubsetSize: model.featureSubsetSize },
    classes: dataset.classes,
    featureColumns: [...FEATURE_COLUMNS],
    accuracy: evalResult.accuracy,
    precision: evalResult.precision,
    recall: evalResult.recall,
    f1: evalResult.f1,
    confusionMatrix: evalResult.confusionMatrix,
    perClass: evalResult.perClass,
  };

  const generated = [
    "/**",
    " * AUTO-GENERATED by ml/train_model.ts — do not edit by hand.",
    " * Regenerate with: bun run ml:train",
    " */",
    'import type { ModelInfo, RandomForestModel } from "./ml/types";',
    "",
    `export const MODEL: RandomForestModel = ${JSON.stringify(model, null, 1)};`,
    "",
    `export const MODEL_INFO: ModelInfo = ${JSON.stringify(info, null, 1)};`,
    "",
  ].join("\n");

  writeFileSync(OUTPUT_PATH, generated, "utf-8");
  console.log("      → model artifact written.\n");

  console.log("═══════════════════════════════════════════════════");
  console.log("  Training complete.");
  console.log(`  Test accuracy: ${pct(evalResult.accuracy)}`);
  console.log("═══════════════════════════════════════════════════");
}

main();
